server/src/main/java/org/elasticsearch/index/mapper/SortedSetDocValuesSyntheticFieldLoaderLayer.java

```diff
@@ -133,6 +149,98 @@ public void write(XContentBuilder b) throws IOException {
         }
     }
 
+    class OffsetDocValuesLoader implements DocValuesLoader, DocValuesFieldValues {
```

Comment:
I think this would be better structured if this was a completely new layer (it looks like all existing functionality is overridden). Then if `offsetsFullPath` is not null you use the new layer, and if it is you continue to use `SortedSetDocValuesSyntheticFieldLoaderLayer`.

Reply:
Agreed, a new layer makes more sense and is cleaner.